### ISSUE TITLE
Temporary Memory Allocator [C/C]

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8674]
Temporary Memory Allocator [C/C++]<br>
<br>
https://ift.tt/iUnR5Cq<br>
<br>
Austin Morlan<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>Jan 31, 2022</p>
<h1>Temporary Memory Allocator [C/C++]</h1>
<hr>
<p>While programming you often need some memory to do something but you don&rsquo;t know at compile-time how much you need. Instead you have to use <strong>malloc</strong> from C or <strong>new</strong> from C++ (or more commonly <strong>std::vector</strong>) to request a certain amount of memory to be allocated for you on the heap. That&rsquo;s just how it works.</p>
<p>You might not question this sort of pattern while doing normal programming but once you start getting into game development you&rsquo;ll quickly learn these sorts of dynamic memory allocations that occur at runtime can have performance implications.</p>
<p>There is an interesting alternative to some dynamic memory allocation that I want to discuss: a temporary memory allocator.</p>
<h2>Examples of Temporary Heap Usage</h2>
<hr>
<p>Consider the following pattern when working with the Vulkan API:</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span><span>14
</span><span>15
</span><span>16
</span><span>17
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>// C
</span></span></span><span><span><span>uint32_t</span> <span>layer_count</span> <span>=</span> <span>0</span><span>;</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>NULL</span><span>);</span>
</span></span><span><span><span>VkLayerProperties</span><span>*</span> <span>layer_properties</span> <span>=</span> <span>malloc</span><span>(</span><span>sizeof</span><span>(</span><span>*</span><span>layer_properties</span><span>)</span> <span>*</span> <span>layer_count</span><span>);</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>layer_properties</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// C++
</span></span></span><span><span><span>uint32_t</span> <span>layer_count</span> <span>=</span> <span>0</span><span>;</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>NULL</span><span>);</span>
</span></span><span><span><span>VkLayerProperties</span><span>*</span> <span>layer_properties</span> <span>=</span> <span>new</span> <span>VkLayerProperties</span><span>[</span><span>layer_count</span><span>];</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>layer_properties</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// C++ (alternative)
</span></span></span><span><span><span>uint32_t</span> <span>layer_count</span> <span>=</span> <span>0</span><span>;</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>NULL</span><span>);</span>
</span></span><span><span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>VkLayerProperties</span><span>&gt;</span> <span>layer_properties</span><span>(</span><span>layer_count</span><span>);</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>layer_properties</span><span>.</span><span>data</span><span>());</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>This pattern repeats itself many times. You first query to see how many of something there are, you allocate some memory to hold the objects, and then you fill the memory.</p>
<p>Another example is when you need to load a file into memory so that you can then process it into a different form at which point you no longer need the memory. Something like a GPU shader or a model loaded from an OBJ file.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>FILE</span><span>*</span> <span>fp</span> <span>=</span> <span>fopen</span><span>(</span><span>path</span><span>,</span> <span>"r"</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>fseek</span><span>(</span><span>fp</span><span>,</span> <span>0</span><span>,</span> <span>SEEK_END</span><span>);</span>
</span></span><span><span><span>size_t</span> <span>file_size</span> <span>=</span> <span>ftell</span><span>(</span><span>fp</span><span>);</span>
</span></span><span><span><span>rewind</span><span>(</span><span>fp</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>char</span><span>*</span> <span>buf</span> <span>=</span> <span>malloc</span><span>(</span><span>file_size</span><span>);</span>
</span></span><span><span><span>fread</span><span>(</span><span>buf</span><span>,</span> <span>sizeof</span><span>(</span><span>*</span><span>buf</span><span>),</span> <span>file_size</span><span>,</span> <span>fp</span><span>);</span>
</span></span><span><span><span>fclose</span><span>(</span><span>fp</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// Create a shader from the contents of the file buf...
</span></span></span><span><span>
</span></span><span><span><span>free</span><span>(</span><span>buf</span><span>);</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<h2>Some Problems with the Heap</h2>
<hr>
<p>That&rsquo;s all well and good, but now you have a pointer to some memory that was allocated dynamically from the heap, and you must remember to free it.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span>1
</span><span>2
</span><span>3
</span><span>4
</span><span>5
</span><span>6
</span><span>7
</span><span>8
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>// C
</span></span></span><span><span><span>free</span><span>(</span><span>layer_properties</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// C++
</span></span></span><span><span><span>delete</span><span>[]</span> <span>layer_properties</span><span>;</span>
</span></span><span><span>
</span></span><span><span><span>// C++ (alternative)
</span></span></span><span><span><span>// Memory is freed when the std::vector goes out of scope
</span></span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>That allocation/free pair is what causes memory leaks: you allocate some memory with <strong>malloc</strong> or <strong>new</strong> but you forget to then use <strong>free</strong> or <strong>delete</strong>. However, if you use C++&rsquo;s <strong>std::vector</strong> then the memory will be freed for you when the vector goes out of scope.</p>
<p>If you do remember to free the memory but then you accidentally try to use the pointer afterward, you&rsquo;ll get a segmentation fault. This sort of thing is what <strong>Resource Acquisition Is Initialization (RAII)</strong> and smart pointers in C++ are supposed to help with. If you allocate memory in the constructor and free it in the destructor then you don&rsquo;t have to worry about memory leaking. Or, if you use a smart pointer, the memory will be freed automatically when the pointer goes out of scope.</p>
<p>There are performance implications as well. <strong>malloc</strong> and <strong>new</strong> are written to be general-purpose and work for all cases all the time, and that generality will have some overhead. Furthermore, each call to <strong>malloc</strong> or <strong>new</strong> (which <strong>std::vector</strong> is doing under the hood) requires a context switch from user mode to kernel mode and then a switch back from kernel mode to user mode. These context switches may not have much of an impact when done rarely but the performance cost adds up when you&rsquo;re doing it many times per frame.</p>
<p>Finally, frequently allocating and freeing memory from the heap can cause memory fragmentation and eventual failure. As you request little chunks of memory from the OS and then free them, the heap space gets full of holes. At some point it&rsquo;s possible to request a chunk of memory and it to fail because the operating system can&rsquo;t find a contiguous chunk big enough. Even though there might be enough free space in total, there isn&rsquo;t enough free space in one contiguous chunk. If that happens then the calls to <strong>malloc</strong> and <strong>new</strong> will return a null pointer and your game will crash when you attempt to use it.</p>
<p>These sorts of issues are more applicable to games than to other sorts of software because games need to run with high performance and they need to be able to run for many hours without crashing.</p>
<h2>Alternative #1: Static Arrays</h2>
<hr>
<p>You can often get away with just allocating a static array big enough to handle the largest amount of data you might ever need.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span>1
</span><span>2
</span><span>3
</span><span>4
</span><span>5
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>uint32_t</span> <span>layer_count</span> <span>=</span> <span>0</span><span>;</span>
</span></span><span><span><span>VkLayerProperties</span> <span>layer_properties</span><span>[</span><span>32</span><span>];</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>NULL</span><span>);</span>
</span></span><span><span><span>assert</span><span>(</span><span>layer_count</span> <span>&lt;</span> <span>32</span><span>);</span>
</span></span><span><span><span>vkEnumerateInstanceLayerProperties</span><span>(</span><span>&amp;</span><span>layer_count</span><span>,</span> <span>layer_properties</span><span>);</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>In this case we make the assumption that there will never be more than 32 layer properties and so we can just use an array allocated on the stack. We can use an <strong>assert</strong> to check our assumption.</p>
<p>This is the simplest approach but it has the side effect of wasted memory. If you know that on average you&rsquo;ll need eight of some item but every so often you need one thousand, then you need to have an array that can hold one thousand.</p>
<p>You also might need a larger array than your stack size allows which will cause a crash.</p>
<p>But it can sometimes still be a good choice.</p>
<h2>Alternative #2: Temporary Memory Allocator</h2>
<hr>
<p>This is my favorite method because of its elegance and simplicity.</p>
<p>You allocate a large pool of memory from the heap at startup and then you allocate from that as needed. The memory allocation happens only once, removing the costly calls to the OS, but you&rsquo;re still able to request arbitrary amounts of memory as if you were using <strong>malloc</strong> or <strong>new</strong>. But it uses simple pointer arithmetic and a stack which is a lot simpler than more complicated general-purpose memory allocators.</p>
<p>This implementation assumes there is only a single thread using the temp memory, and it doesn't consider memory alignment. You may wish to only hand out memory addresses that are aligned for SIMD access, for example.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span>1
</span><span>2
</span><span>3
</span><span>4
</span><span>5
</span><span>6
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>typedef</span> <span>struct</span> <span>{</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>current</span><span>;</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>end</span><span>;</span>
</span></span><span><span>      <span>int</span> <span>index</span><span>;</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>stack</span><span>[</span><span>MAX_STACK_COUNT</span><span>];</span>
</span></span><span><span><span>}</span> <span>temp_mem_t</span><span>;</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>We define a simple structure that is nothing more than some pointers and an index. Its general usage looks something like this:</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span><span>14
</span><span>15
</span><span>16
</span><span>17
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>// Request 16 bytes of temporary memory (once at the beginning of the program)
</span></span></span><span><span><span>temp_mem_t</span> <span>mem</span> <span>=</span> <span>create_temp_mem</span><span>(</span><span>16</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// Begin a scoped temp memory frame
</span></span></span><span><span><span>begin_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// Request as needed
</span></span></span><span><span><span>uint8_t</span><span>*</span> <span>data0</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>4</span><span>);</span>
</span></span><span><span><span>uint8_t</span><span>*</span> <span>data1</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>1</span><span>);</span>
</span></span><span><span><span>uint8_t</span><span>*</span> <span>data2</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>6</span><span>);</span>
</span></span><span><span><span>uint8_t</span><span>*</span> <span>data3</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>4</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// End the scoped frame
</span></span></span><span><span><span>end_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>// Free the memory at the very end of the program
</span></span></span><span><span><span>temp_mem_destroy</span><span>(</span><span>&amp;</span><span>mem</span><span>);</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>We first need to create a <strong>temp_mem_t</strong> object which we can then use throughout the lifetime of our game. Before we use any temp memory we first call <strong>begin_temp_mem</strong> and then when we&rsquo;re done we call <strong>end_temp_mem</strong>. At the end of our game during shutdown we can destroy it.</p>
<h3>Stepping Through</h3>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>temp_mem_t</span> <span>create_temp_mem</span><span>(</span><span>size_t</span> <span>size</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>temp_mem_t</span> <span>temp_mem</span> <span>=</span> <span>{</span><span>0</span><span>};</span>
</span></span><span><span>      <span>temp_mem</span><span>.</span><span>current</span> <span>=</span> <span>malloc</span><span>(</span><span>size</span><span>);</span>
</span></span><span><span>      <span>temp_mem</span><span>.</span><span>end</span> <span>=</span> <span>temp_mem</span><span>.</span><span>current</span> <span>+</span> <span>size</span><span>;</span>
</span></span><span><span>
</span></span><span><span>      <span>// Make sure we were able to get enough memory
</span></span></span><span><span>  <span>assert</span><span>(</span><span>temp_mem</span><span>.</span><span>current</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>return</span> <span>temp_mem</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>Initialization is simple. We use <strong>malloc</strong> to request a large chunk of size <strong>size</strong> from the OS and we store the address of that chunk in the <strong>current</strong> variable. We also cache an <strong>end</strong> pointer which points to the end of the requested chunk. We can use the <strong>end</strong> pointer to check and make sure we never exceed the amount of memory that we initially allocated.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span><span>14
</span><span>15
</span><span>16
</span><span>17
</span><span>18
</span><span>19
</span><span>20
</span><span>21
</span><span>22
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>void</span> <span>begin_temp_mem</span><span>(</span><span>temp_mem_t</span><span>*</span> <span>mem</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>// Make sure that we haven't nested too deeply
</span></span></span><span><span>  <span>assert</span><span>(</span><span>mem</span><span>-&gt;</span><span>index</span> <span>&lt;</span> <span>MAX_STACK_COUNT</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>// Save our previous memory location
</span></span></span><span><span>  <span>mem</span><span>-&gt;</span><span>stack</span><span>[</span><span>mem</span><span>-&gt;</span><span>index</span><span>]</span> <span>=</span> <span>mem</span><span>-&gt;</span><span>current</span><span>;</span>
</span></span><span><span>      <span>mem</span><span>-&gt;</span><span>index</span><span>++</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>end_temp_mem</span><span>(</span><span>temp_mem_t</span><span>*</span> <span>mem</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>// Make sure we haven't called end() more times than begin()
</span></span></span><span><span>  <span>assert</span><span>(</span><span>mem</span><span>-&gt;</span><span>index</span> <span>&gt;</span> <span>0</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>// Restore our state to the previous memory location
</span></span></span><span><span>  <span>mem</span><span>-&gt;</span><span>index</span><span>--</span><span>;</span>
</span></span><span><span>      <span>mem</span><span>-&gt;</span><span>current</span> <span>=</span> <span>mem</span><span>-&gt;</span><span>stack</span><span>[</span><span>mem</span><span>-&gt;</span><span>index</span><span>];</span>
</span></span><span><span>
</span></span><span><span>      <span>// This isn't necessary but it makes things clearer when stepping through a debugger
</span></span></span><span><span>  <span>mem</span><span>-&gt;</span><span>stack</span><span>[</span><span>mem</span><span>-&gt;</span><span>index</span><span>]</span> <span>=</span> <span>NULL</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>A scope begins by pushing the old <strong>current</strong> pointer to the stack, and a scope ends by popping a pointer from the stack back into the <strong>current</strong> pointer.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>void</span><span>*</span> <span>get_temp_mem</span><span>(</span><span>temp_mem_t</span><span>*</span> <span>mem</span><span>,</span> <span>size_t</span> <span>size</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>buf</span> <span>=</span> <span>(</span><span>uint8_t</span><span>*</span><span>)</span><span>mem</span><span>-&gt;</span><span>current</span> <span>+</span> <span>size</span><span>;</span>
</span></span><span><span>
</span></span><span><span>      <span>// Make sure that we aren't requesting more memory than we have available
</span></span></span><span><span>  <span>assert</span><span>(</span><span>buf</span> <span>&lt;</span> <span>mem</span><span>-&gt;</span><span>end</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>// Save where we were
</span></span></span><span><span>  <span>void</span><span>*</span> <span>old</span> <span>=</span> <span>mem</span><span>-&gt;</span><span>current</span><span>;</span>
</span></span><span><span>      <span>mem</span><span>-&gt;</span><span>current</span> <span>=</span> <span>buf</span><span>;</span>
</span></span><span><span>
</span></span><span><span>      <span>return</span> <span>old</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>Requesting new memory is a matter of moving the <strong>current</strong> pointer to accomodate the new size and returning the old pointer which can now be filled out safely.</p>
<p>In our example, we requested a temp memory buffer of 16 bytes. In reality you&rsquo;d likely want something more like a megabyte or gigabyte (depending on your application), but 16 bytes is nice for demonstration purposes. We also assume that <strong>malloc</strong> returns an address of 0x00, again for demonstration purposes.</p>
<p>At this point we can request memory from the temporary pool and it will return the <strong>current</strong> pointer and then increment it by the amount of bytes requested. That way any subsequent requests won&rsquo;t clobber the data that&rsquo;s already been given out.</p>
<h3>The Stack</h3>
<p>What happens if you&rsquo;ve requested some temporary memory (so you&rsquo;re inside of a begin/end block) and then you call a function which itself requests some temp memory? You&rsquo;d clobber the other function&rsquo;s data. You need a way to keep track of the pointers so that you don&rsquo;t overwrite anything and so you can return back to where you were.</p>
<p>Each call to <strong>begin_temp_mem</strong> pushes the previous address onto the stack and each subsequent call to <strong>end_temp_mem</strong> pops that value off the stack into <strong>current</strong>.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span> 10
</span><span> 11
</span><span> 12
</span><span> 13
</span><span> 14
</span><span> 15
</span><span> 16
</span><span> 17
</span><span> 18
</span><span> 19
</span><span> 20
</span><span> 21
</span><span> 22
</span><span> 23
</span><span> 24
</span><span> 25
</span><span> 26
</span><span> 27
</span><span> 28
</span><span> 29
</span><span> 30
</span><span> 31
</span><span> 32
</span><span> 33
</span><span> 34
</span><span> 35
</span><span> 36
</span><span> 37
</span><span> 38
</span><span> 39
</span><span> 40
</span><span> 41
</span><span> 42
</span><span> 43
</span><span> 44
</span><span> 45
</span><span> 46
</span><span> 47
</span><span> 48
</span><span> 49
</span><span> 50
</span><span> 51
</span><span> 52
</span><span> 53
</span><span> 54
</span><span> 55
</span><span> 56
</span><span> 57
</span><span> 58
</span><span> 59
</span><span> 60
</span><span> 61
</span><span> 62
</span><span> 63
</span><span> 64
</span><span> 65
</span><span> 66
</span><span> 67
</span><span> 68
</span><span> 69
</span><span> 70
</span><span> 71
</span><span> 72
</span><span> 73
</span><span> 74
</span><span> 75
</span><span> 76
</span><span> 77
</span><span> 78
</span><span> 79
</span><span> 80
</span><span> 81
</span><span> 82
</span><span> 83
</span><span> 84
</span><span> 85
</span><span> 86
</span><span> 87
</span><span> 88
</span><span> 89
</span><span> 90
</span><span> 91
</span><span> 92
</span><span> 93
</span><span> 94
</span><span> 95
</span><span> 96
</span><span> 97
</span><span> 98
</span><span> 99
</span><span>100
</span><span>101
</span><span>102
</span><span>103
</span><span>104
</span><span>105
</span><span>106
</span><span>107
</span><span>108
</span><span>109
</span><span>110
</span><span>111
</span><span>112
</span><span>113
</span><span>114
</span><span>115
</span><span>116
</span><span>117
</span><span>118
</span><span>119
</span><span>120
</span><span>121
</span><span>122
</span><span>123
</span><span>124
</span><span>125
</span><span>126
</span><span>127
</span><span>128
</span><span>129
</span><span>130
</span><span>131
</span><span>132
</span><span>133
</span><span>134
</span><span>135
</span><span>136
</span><span>137
</span><span>138
</span><span>139
</span><span>140
</span><span>141
</span><span>142
</span><span>143
</span><span>144
</span><span>145
</span><span>146
</span><span>147
</span><span>148
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>void</span> <span>foo</span><span>(</span><span>temp_mem_t</span><span>*</span> <span>mem</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 0
</span></span></span><span><span><span> mem.current: 0x00
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: NULL &lt;-- index
</span></span></span><span><span><span> 1: NULL
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>begin_temp_mem</span><span>(</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 1
</span></span></span><span><span><span> mem.current: 0x00
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: NULL &lt;-- index
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>// data0 = 0x00
</span></span></span><span><span>  <span>uint8_t</span><span>*</span> <span>data0</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>mem</span><span>,</span> <span>4</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 1
</span></span></span><span><span><span> mem.current: 0x04
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: NULL &lt;-- index
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>// See bar() definition below
</span></span></span><span><span>  <span>bar</span><span>(</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>// data1 = 0x04
</span></span></span><span><span>  <span>uint8_t</span><span>*</span> <span>data1</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>mem</span><span>,</span> <span>1</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 1
</span></span></span><span><span><span> mem.current: 0x05
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: NULL &lt;-- index
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>end_temp_mem</span><span>(</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 0
</span></span></span><span><span><span> mem.current: 0x00
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: NULL &lt;-- index
</span></span></span><span><span><span> 1: NULL
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>bar</span><span>(</span><span>temp_mem_t</span><span>*</span> <span>mem</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 1
</span></span></span><span><span><span> mem.current: 0x04
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: NULL &lt;-- index
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>begin_temp_mem</span><span>(</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 2
</span></span></span><span><span><span> mem.current: 0x04
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: 0x04
</span></span></span><span><span><span> 2: NULL &lt;-- index
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>// data1 = 0x04
</span></span></span><span><span>  <span>uint8_t</span><span>*</span> <span>data0</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>mem</span><span>,</span> <span>2</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 2
</span></span></span><span><span><span> mem.current: 0x06
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: 0x04
</span></span></span><span><span><span> 2: NULL &lt;-- index
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>// data1 = 0x06
</span></span></span><span><span>  <span>uint8_t</span><span>*</span> <span>data1</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>mem</span><span>,</span> <span>2</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 2
</span></span></span><span><span><span> mem.current: 0x08
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: 0x04
</span></span></span><span><span><span> 2: NULL &lt;-- index
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span>
</span></span><span><span>      <span>end_temp_mem</span><span>(</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>/*
</span></span></span><span><span><span> mem.index: 1
</span></span></span><span><span><span> mem.current: 0x04
</span></span></span><span><span><span>
</span></span></span><span><span><span> STACK
</span></span></span><span><span><span> -------
</span></span></span><span><span><span> 0: 0x00
</span></span></span><span><span><span> 1: NULL &lt;-- index
</span></span></span><span><span><span> 2: NULL
</span></span></span><span><span><span> 3: NULL
</span></span></span><span><span><span> */</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<h3>Managing Scope</h3>
<p>A flaw with this approach is that you must remember to call <strong>end_temp_mem</strong>. There&rsquo;s a cool trick we can use to create some macros that will catch that error for us.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span><span>14
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>#define BEGIN_TEMP_MEM(mem) \
</span></span></span><span><span><span>do { \
</span></span></span><span><span><span> goto temp_label_end; \
</span></span></span><span><span><span>temp_label_begin: \
</span></span></span><span><span><span> begin_temp_mem(mem);\
</span></span></span><span><span><span>} while (0)
</span></span></span><span><span>
</span></span><span><span><span>#define END_TEMP_MEM(mem) \
</span></span></span><span><span><span>do { \
</span></span></span><span><span><span> end_temp_mem(mem); \
</span></span></span><span><span><span> break; \
</span></span></span><span><span><span>temp_label_end: \
</span></span></span><span><span><span> goto temp_label_begin; \
</span></span></span><span><span><span>} while (0)
</span></span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>The <strong>do while</strong> part is just macro magic to ensure we can put our macros inside of another block. The goto labels ensure that if we use one with the other we&rsquo;ll have a compile-time error as <strong>temp_label_begin</strong> won&rsquo;t be defined without <strong>BEGIN_TEMP_MEM</strong> and <strong>temp_label_end</strong> won&rsquo;t be defined without <strong>END_TEMP_MEM</strong>. The only catch is that we can only have one temp memory region per function.</p>
<p>We can then use them like this:</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span><span>14
</span><span>15
</span><span>16
</span><span>17
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>int</span> <span>main</span><span>(</span><span>void</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>temp_mem_t</span> <span>mem</span> <span>=</span> <span>create_temp_mem</span><span>(</span><span>16</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>BEGIN_TEMP_MEM</span><span>(</span><span>&amp;</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>uint8_t</span><span>*</span> <span>data0</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>4</span><span>);</span>
</span></span><span><span>      <span>uint8_t</span><span>*</span> <span>data1</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>1</span><span>);</span>
</span></span><span><span>      <span>uint8_t</span><span>*</span> <span>data2</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>6</span><span>);</span>
</span></span><span><span>      <span>uint8_t</span><span>*</span> <span>data3</span> <span>=</span> <span>get_temp_mem</span><span>(</span><span>&amp;</span><span>mem</span><span>,</span> <span>4</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>END_TEMP_MEM</span><span>(</span><span>&amp;</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>temp_mem_destroy</span><span>(</span><span>&amp;</span><span>mem</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>return</span> <span>0</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<p>If using C++ we can help prevent this by creating a separate class that calls <strong>begin_temp_mem</strong> in its constructor and <strong>end_temp_mem</strong> in its destructor. Then we just instantiate the class at the beginning of our tempory memory scope and when falling out of scope the destructor will be called automatically which will call <strong>end_temp_mem</strong> for us.</p>
<p>Personally I prefer the C approach of seeing the explicit calls to <strong>begin_temp_mem</strong> and <strong>end_temp_mem</strong>, but I recognize the danger of forgetting to call one or the other.</p>
<h3>A Global Object</h3>
<p>You might also notice that passing around a <strong>temp_mem_t</strong> object to every function is a pain. Similar to something like a logger, it would be better to create some global object somewhere that everyone can access. You might disagree with this philosophically (&ldquo;globals are bad&rdquo;) but it works. You could also create a static singleton class that holds a reference to the temporary memory but not everyone is a fan of singletons either.</p>
<p>You can handle it however you want. Since I&rsquo;m the only person working on my personal projects, I opt for a single <strong>temp_mem_t</strong> that is static to my <strong>memory.c</strong> file. I can just call the relevant functions without needing to worry about the actual <strong>temp_mem_t</strong> object.</p>
<h3>In Practice</h3>
<p>Here is how I would do it if you just wanted a single <strong>temp_mem_t</strong> that is static to the allocator file.</p>
<div>
<div>
<table>
<tr>
<td>
<pre tabindex="0"><code><span> 1
</span><span> 2
</span><span> 3
</span><span> 4
</span><span> 5
</span><span> 6
</span><span> 7
</span><span> 8
</span><span> 9
</span><span>10
</span><span>11
</span><span>12
</span><span>13
</span><span>14
</span><span>15
</span><span>16
</span><span>17
</span><span>18
</span><span>19
</span><span>20
</span><span>21
</span><span>22
</span><span>23
</span><span>24
</span><span>25
</span><span>26
</span><span>27
</span><span>28
</span><span>29
</span><span>30
</span><span>31
</span><span>32
</span><span>33
</span><span>34
</span><span>35
</span><span>36
</span><span>37
</span><span>38
</span><span>39
</span><span>40
</span><span>41
</span><span>42
</span><span>43
</span><span>44
</span><span>45
</span><span>46
</span><span>47
</span><span>48
</span><span>49
</span><span>50
</span><span>51
</span><span>52
</span><span>53
</span><span>54
</span><span>55
</span><span>56
</span><span>57
</span><span>58
</span></code></pre>
</td>
<td>
<pre tabindex="0"><code><span><span><span>/* memory.c */</span>
</span></span><span><span>
</span></span><span><span><span>#define TEMP_MEM_SIZE (1U * 1024U * 1024U)
</span></span></span><span><span><span>#define TEMP_MEM_STACK_LENGTH (8)
</span></span></span><span><span>
</span></span><span><span><span>static</span> <span>struct</span> <span>{</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>current</span><span>;</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>end</span><span>;</span>
</span></span><span><span>      <span>size_t</span> <span>index</span><span>;</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>stack</span><span>[</span><span>TEMP_MEM_STACK_LENGTH</span><span>];</span>
</span></span><span><span><span>}</span> <span>s_temp_mem</span><span>;</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>allocator_startup</span><span>(</span><span>void</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>current</span> <span>=</span> <span>malloc</span><span>(</span><span>TEMP_MEM_SIZE</span><span>);</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>end</span> <span>=</span> <span>(</span><span>uint8_t</span><span>*</span><span>)</span><span>s_temp_mem</span><span>.</span><span>current</span> <span>+</span> <span>TEMP_MEM_SIZE</span><span>;</span>
</span></span><span><span>
</span></span><span><span>      <span>assert</span><span>(</span><span>s_temp_mem</span><span>.</span><span>current</span><span>);</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>allocator_shutdown</span><span>(</span><span>void</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>free</span><span>(</span><span>s_temp_mem</span><span>.</span><span>current</span><span>);</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>current</span> <span>=</span> <span>NULL</span><span>;</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>end</span> <span>=</span> <span>NULL</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>begin_temp_mem</span><span>(</span><span>void</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>assert</span><span>(</span><span>s_temp_mem</span><span>.</span><span>index</span> <span>&lt;</span> <span>TEMP_MEM_STACK_LENGTH</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>stack</span><span>[</span><span>s_temp_mem</span><span>.</span><span>index</span><span>]</span> <span>=</span> <span>s_temp_mem</span><span>.</span><span>current</span><span>;</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>index</span><span>++</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>end_temp_mem</span><span>(</span><span>void</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>assert</span><span>(</span><span>s_temp_mem</span><span>.</span><span>index</span> <span>&gt;</span> <span>0</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>index</span><span>--</span><span>;</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>current</span> <span>=</span> <span>s_temp_mem</span><span>.</span><span>stack</span><span>[</span><span>s_temp_mem</span><span>.</span><span>index</span><span>];</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>stack</span><span>[</span><span>s_temp_mem</span><span>.</span><span>index</span><span>]</span> <span>=</span> <span>NULL</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span><span>*</span> <span>get_temp_mem</span><span>(</span><span>size_t</span> <span>size</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span>      <span>void</span><span>*</span> <span>buf</span> <span>=</span> <span>(</span><span>uint8_t</span><span>*</span><span>)</span><span>s_temp_mem</span><span>.</span><span>current</span> <span>+</span> <span>size</span><span>;</span>
</span></span><span><span>
</span></span><span><span>      <span>assert</span><span>(</span><span>buf</span> <span>&lt;</span> <span>s_temp_mem</span><span>.</span><span>end</span><span>);</span>
</span></span><span><span>
</span></span><span><span>      <span>void</span><span>*</span> <span>old</span> <span>=</span> <span>s_temp_mem</span><span>.</span><span>current</span><span>;</span>
</span></span><span><span>      <span>s_temp_mem</span><span>.</span><span>current</span> <span>=</span> <span>buf</span><span>;</span>
</span></span><span><span>
</span></span><span><span>      <span>return</span> <span>old</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>#undef TEMP_MEM_SIZE
</span></span></span><span><span><span>#undef TEMP_MEM_STACK_LENGTH
</span></span></span></code></pre>
</td>
</tr>
</table>
</div>
</div>
<h2>Conclusion</h2>
<hr>
<p>Hopefully you find the concept of a large chunk of temporary memory to be as compelling and elegant as I do.</p>
<p>There is another side to the dynamic memory coin which is when you need memory from the heap but with a lifetime that is larger than a single scope. That requires a custom memory allocator which again allocates a large chunk during game startup but it&rsquo;s responsible for handing out memory, keeping a free list, rearranging to remove holes, etc. That&rsquo;s a topic for a different day.</p>
<h2>Further Reading</h2>
<hr>
<h2>Discussion</h2>
<hr>
<p><a href="https://old.reddit.com/r/austinmorlan/comments/v4cnsa/temporary_memory_allocator_cc/">Discuss here</a></p>
<p>Last Edited: Jul 09, 2022</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Austin Morlan<br>
<br>
September 1, 2022 at 02:44PM
